### PR TITLE
Add amulet of chemistry to the item charges plugin

### LINKED ISSUE #9246: Amulet of Chemistry/similar item replace overlay
**Is your feature request related to a problem? Please describe.**
When I'm making potions I'm always having to check my equipment spot to see if the amulet has run out. This is because when I'm making prayer potions for example if the amulet breaks on the 14th slot it's hard to tell whether the amulet needs replacing without checking equipment, because the message flies by the chatbox.
 
**Describe the solution you'd like**
An overlay telling me I need to replace the amulet. I know other amulets/rings have similar effects, so it would be nice if they could get similar treatment too.

### PR DESCRIPTION
**Pull request by Barragek0**
Adds the amulet of chemistry to the item charges plugin
tScKelI
RR26biC

Closes #9246

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java [hunk 1]
```diff
@@ -72,9 +72,16 @@
 	private static final Pattern BINDING_USED_PATTERN = Pattern.compile(
 		"You bind the temple's power into (mud|lava|steam|dust|smoke|mist) runes\\.");
 	private static final String BINDING_BREAK_TEXT = "Your Binding necklace has disintegrated.";
+	private static final Pattern CHEMISTRY_CHECK_PATTERN = Pattern.compile(
+		"Your amulet of chemistry helps you create a 4-dose potion. It has (\\d+) charges? left\\.");
+	private static final Pattern CHEMISTRY_CHECK2_PATTERN = Pattern.compile(
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java]: Use a better variable name, such as `CHEMISTRY_ONE_CHARGE_PATTERN`.

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java [hunk 2]
```diff
@@ -214,6 +229,25 @@ else if (bindingNecklaceCheckMatcher.find())
 
 				updateBindingNecklaceCharges(charges);
 			}
+			else if (chemistryUsedMatcher.find())
+			{
+				if (config.chemistryNotification())
+				{
+					notifier.notify("Your amulet of chemistry has been destroyed.");
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java]:
```suggestion
					notifier.notify("Your amulet of chemistry has crumbled to dust.");
```

**Comment by Nightfirecat** [comment]
Also, is there a chat message or any kind of indication when you manually break an amulet of chemistry? If so, it does not appear to be handled in this PR.

**Comment by Barragek0** [comment]
> 
> 
> Also, is there a chat message or any kind of indication when you manually break an amulet of chemistry? If so, it does not appear to be handled in this PR.

I don't think there's any way to manually break one. It is handled when using the final charge to make a 4 dose potion if that's what you mean via CHEMISTRY_USED_PATTERN

**Comment by Hydrox6** [comment]
Screenshot_2019-05-16_00-20-39
When breaking, it uses widgets, not a chat message

**Comment by Barragek0** [comment]
> 
> 
> Screenshot_2019-05-16_00-20-39
> When breaking, it uses widgets, not a chat message

Yep, forgot that was an option, did edit the comment before.

**Comment by Nightfirecat** [comment]
Then it should be handled like in #7969

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java [hunk 3]
```diff
@@ -311,15 +397,23 @@ private void checkDestroyWidget()
 		{
 			return;
 		}
-
 		switch (widgetDestroyItemName.getText())
 		{
 			case "Binding necklace":
 				updateBindingNecklaceCharges(MAX_BINDING_CHARGES);
 				break;
-			case "Dodgy necklace":
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java]: ??

Comment (Barragek0) [reply runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java]: Checkdestroywidget doesnt work since it has the 'Break' option

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java]: It should be in separate commit then. Also, I do not see any bug report for this.

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java [hunk 4]
```diff
@@ -263,6 +306,32 @@ private void onScriptCallbackEvent(ScriptCallbackEvent event)
 		}
 	}
 
+	@Subscribe
+	private void onWidgetLoaded(WidgetLoaded event)
+	{
+		if (event.getGroupId() != 193)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java]: You should add this groupId to some of the widget constant files

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java [hunk 5]
```diff
@@ -263,6 +306,32 @@ private void onScriptCallbackEvent(ScriptCallbackEvent event)
...
+		{
+			return;
+		}
+		waitingForDialogue = false;
+		final String text = client.getWidget(WidgetInfo.DIALOG_SPRITE_TEXT).getText();
+		if (!text.contains("amulet shatters"))
+		{
+			return;
+		}
+		checkBreakWidget(text);
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java]: new line above

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java [hunk 6]
```diff
@@ -263,6 +306,32 @@ private void onScriptCallbackEvent(ScriptCallbackEvent event)
...
+	}
+
+	@Subscribe
+	private void onGameTick(GameTick event)
+	{
+		if (!waitingForDialogue)
+		{
+			return;
+		}
+		waitingForDialogue = false;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java]: new line above

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java [hunk 7]
```diff
@@ -311,7 +397,6 @@ private void checkDestroyWidget()
 		{
 			return;
 		}
-
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java]: put this new line back

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java [hunk 8]
```diff
@@ -390,7 +487,7 @@ private void removeInfobox(final ItemWithSlot item, final EquipmentInventorySlot
 				return false;
 			}
 
-			final ItemChargeInfobox i = (ItemChargeInfobox)t;
+			final ItemChargeInfobox i = (ItemChargeInfobox) t;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java]: revert this change

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java [hunk 9]
```diff
@@ -263,6 +307,34 @@ private void onScriptCallbackEvent(ScriptCallbackEvent event)
...
 	}
 
+	@Subscribe
+	private void onWidgetLoaded(WidgetLoaded event)
+	{
+		if (event.getGroupId() != WidgetID.DIALOG_SPRITE_GROUP_ID)
+		{
+			return;
+		}
+		waitingForDialogue = true;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java]: add empty line above

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java [hunk 10]
```diff
@@ -263,6 +307,34 @@ private void onScriptCallbackEvent(ScriptCallbackEvent event)
...
+	@Subscribe
+	private void onGameTick(GameTick event)
+	{
+		if (!waitingForDialogue)
+		{
+			return;
+		}
+
+		waitingForDialogue = false;
+		final String text = client.getWidget(WidgetInfo.DIALOG_SPRITE_TEXT).getText();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java]: add empty line above

**Comment by Adam-** [comment]
Does `destroyOnOpKey` not fire for destroying this like the other jewelry we have?

**Comment by Barragek0** [comment]
> 
> 
> Does `destroyOnOpKey` not fire for destroying this like the other jewelry we have?

Nope, it has the empty option instead of destroy.
